ls1intum/Artemis
# [Chore / Fix] Lazy load brace module

### Checklist

### Motivation and Context



The brace module for the ace editor is loaded directly into the main Artemis bundle, which increases the bundle size significantly. This PR fixes this so that brace is lazily loaded.

### Description


The direct imports causing the problem were:
- ArtemisAppModule -> TeamModule -> ProgrammingAssessmentModule -> CodeEditorModule -> ... -> brace
- ArtemisAppModule -> CoursesModule -> StudentQuestionsModule -> MarkdownEditorModule -> ... -> brace

The TeamModule did use the ProgrammingAssessmentManualResultButton which was part of the ProgrammingAssessmentModule. Since no other Component from this module was used, i broke the chain by moving the ProgrammingAssessmentManualResultButton into the ResultModule and updating all dependencies accordingly.
The new import chains are:
- ArtemisAppModule -> TeamModule -> ResultModule -> ... (no brace)
- (Lazy Loading) -> ... -> ProgrammingAssessmentModule -> ResultModule / CodeEditorModule -> ... -> brace

The StudentQuestionsModule was used for the CourseExerciseDetails and LectureExerciseDetails - Components. In order to lazy load this I used the module as a child route to both components with an empty path. Therefore instead of inserting the q&a component directly into the details components, i used <router-outlet> to display the component when loaded.

### Steps for Testing


1. Check that the manual result button still works in all places (only for programming exercises with manual assessment! ):
    - in the Exercise Assessment Dashboard (Course-Management -> Assessment Dashboard -> Exercise Dashboard)
    - in the Exercise Scores Page (Course-Management -> Exercises -> Scores)
    - in the Team Participation Table (Course-Management -> Exercises -> Teams -> Click on a Team Id)
3. Check that the question and answer component works in all places:
    - in the Course Exercise Details Page (Courses -> Click on a Course -> Click on an Exercise)
    - in the Course Lecture Details Page (Courses -> Click on a Course -> Switch to Lectures Tab -> Click on a Lecture)

### Test Coverage

No changes regarding test coverage.

### Screenshots



The Q&A Component
<img width="321" alt="Bildschirmfoto 2020-10-25 um 17 11 11" src="https://user-images.githubusercontent.com/13920539/97112534-17599200-16e5-11eb-8852-abde88e173bf.png">

The Manual Result Button
<img width="113" alt="Bildschirmfoto 2020-10-25 um 17 12 37" src="https://user-images.githubusercontent.com/13920539/97112646-c4cca580-16e5-11eb-982f-980761d39a5b.png">
<img width="125" alt="Bildschirmfoto 2020-10-25 um 17 13 36" src="https://user-images.githubusercontent.com/13920539/97112587-6d2e3a00-16e5-11eb-979a-221af01a6355.png">


### Bundle Size

- Before Changes: 2.46MB gzip size
- After Changes: 1.56MB gzip size
